packages/app/obojobo-express/server/models/edit_lock.js

```diff
@@ -0,0 +1,118 @@
+const db = require('../db')
+const editLockExpireMinutes = oboRequire('server/config').general.editLocks.autoExpireMinutes
+
+class EditLock {
+	constructor(props) {
+		// expand all the props onto this object
+		for (const prop in props) {
+			this[prop] = props[prop]
+		}
+	}
+
+	// returns count of all rows deleted
+	static deleteExpiredLocks() {
+		return db.result(
+			`
+			DELETE
+			FROM edit_locks
+			WHERE created_at < now() - interval '${editLockExpireMinutes} minutes'
+				`,
+			null,
+			r => r.rowCount // extract rowCount from query data
+		)
+	}
+
+	// clear a user's locks on a draft
+	// returns true if deleted, false if nothing to delete
+	static deleteByDraftIdAndUser(userId, draftId) {
+		return db
+			.result(
+				`
+			DELETE
+			FROM edit_locks
+			WHERE draft_id = $[draftId]
+			AND user_id = $[userId]
+				`,
+				{ userId, draftId },
+				r => r.rowCount // extract rowCount from query data
+			)
+			.then(rows => {
+				return rows > 0
+			})
+	}
+
+	// get only the most recent active lock for a draft
+	static fetchByDraftId(draftId) {
+		return db
+			.oneOrNone(
+				`
+			SELECT
+				id,
+				user_id AS "userId",
+				draft_id AS "draftId",
+				created_at AS "createdAt"
+			FROM edit_locks
+			WHERE draft_id = $[draftId]
+			AND created_at > now() - interval '${editLockExpireMinutes} minutes'
+			ORDER BY created_at DESC
+			LIMIT 1
+		`,
+				{ draftId }
+			)
+			.then(data => {
+				if (data === null) return
+				return new EditLock(data)
+			})
+	}
+
+	// inserts a lock only when a different user's valid lock doesn't already exist
+	static async create(userId, draftId, contentId) {
+		return db.taskIf(async t => {
+			// get the newest draft contentId for this draft
+			const currentContentId = await t.one(
+				`SELECT
+					drafts.id AS "draftId",
+					drafts_content.id AS "contentId"
+				FROM drafts
+				JOIN drafts_content
+					ON drafts.id = drafts_content.draft_id
+				WHERE drafts.id = $[draftId]
+					AND deleted = FALSE
+				ORDER BY drafts_content.created_at DESC
+				LIMIT 1
+				`,
+				{ draftId }
+			)
+
+			// verify the newest version is the same as the version requested
+			// if it is not - that means the draft has been updated unexpectedly
+			if (currentContentId.contentId !== contentId) {
+				throw Error('Current version of draft does not match requested lock.')
+			}
+
+			return await t
+				.one(
```

Comment:
Changing to `oneOrNone` resolves the error Zach described above, where attempting to edit a locked module throws an error instead of indicating that the module is locked:
```suggestion
				.oneOrNone(
```

Looks like the issue Elli mentioned is also still in play - would it be a significant imposition to clear the lock when the window unloads?

Reply:
Resolved, used Brandon's suggestion but needed a little more integration.